### PR TITLE
Feature/eval

### PR DESCRIPTION
**Pull request by saragee3**
Closes https://github.com/drashland/sinco/issues/31

## Other Notes
Documentation: https://github.com/drashland/website/pull/346 thanks @ebebbington 😎

### CODE REVIEW

**Comment by ebebbington** [comment]
This PR is blocked due to a bug in Deno. it will be fixed in 1.7.0

File: tests/integration/manipulate_page_test.ts
Comment (ebebbington) [inline tests/integration/manipulate_page_test.ts:17]: Shouldn't the. `document.body.children.length` now be one more than `prevBody`? 

Eg

```
const prevBody = ... // 2 children
document.append(p)
const currentBody. = ... // 3 children
```

So that eval should return false right?

File: tests/unit/headless_browser_test.ts
Comment (ebebbington) [inline tests/unit/headless_browser_test.ts:185]: Note to self too: we'll need to add error handling - what if a user types `evaluatePage("vfkjgjgrbj")`

**Review by ebebbington — APPROVED** [review]

**Comment by ebebbington** [comment]
woops didn't mean to approve

File: tests/integration/manipulate_page_test.ts
Comment (saragee3) [reply tests/integration/manipulate_page_test.ts:17]: this is `true`. `prevBody` is one less than `document.body.children` because we captured that val _before_ the `appendChild` call

so `prevBody === document.body.children.length` will return `false` but `prevBody === document.body.children.length - 1` is accurate

File: tests/unit/headless_browser_test.ts
Comment (saragee3) [reply tests/unit/headless_browser_test.ts:185]: if we want `evaluatePage` to act like the console, then `evaluatePage("vfkjgjgrbj")` will return `"vfkjgjgrbj"` . it is a string expression... BUT, we should add tests that throws for things like `evaluatePage(aaa)`

File: tests/integration/manipulate_page_test.ts
Comment (ebebbington) [reply tests/integration/manipulate_page_test.ts:17]: Ohhh i see!! Woooooowwwwwwwwwwwwwwwwwww and i call myself a developer

File: tests/unit/headless_browser_test.ts
Comment (ebebbington) [reply tests/unit/headless_browser_test.ts:185]: I believe that in other API methods, they do`this.checkForError(result)`, ill give it a try and see what happens

Comment (ebebbington) [reply tests/unit/headless_browser_test.ts:185]: Result inside the evaluate page:
```
{
  result: {
    type: "object",
    subtype: "error",
    className: "ReferenceError",
    description: "ReferenceError: fvejfvkjfkj is not defined\n    at <anonymous>:1:1",
    objectId: '{"injectedScriptId":2,"id":1}'
  },
  exceptionDetails: {
    exceptionId: 1,
    text: "Uncaught",
    lineNumber: 0,
    columnNumber: 0,
    scriptId: "12",
    exception: {
      type: "object",
      subtype: "error",
      className: "ReferenceError",
      description: "ReferenceError: fvejfvkjfkj is not defined\n    at <anonymous>:1:1",
      objectId: '{"injectedScriptId":2,"id":2}'
    }
  }
}
```

Comment (ebebbington) [reply tests/unit/headless_browser_test.ts:185]: so like if someone did add invalid input,we'd usually get:

image

**Comment by ebebbington** [comment]
So i've tested with using std 0.84.0, and the error is fixed... but... another error shows...:

```
$ deno test -A tests/unit --config tsconfig.json 
Check file:///sinco/$deno$test.ts
error: TS7053 [ERROR]: Element implicitly has an 'any' type because expression of type 'symbol' can't be used to index type 'ReadableStream<R>'.
  return res.readable[Symbol.asyncIterator]();
         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    at https://deno.land/std@0.84.0/async/pool.ts:45:10

```

**Comment by ebebbington** [comment]
Logged an issue here: https://github.com/denoland/deno/issues/9218

**Comment by ebebbington** [comment]
This is going to be great for element selection, i just came across the problem of wanting to get the text of a h1 element - ie this eval feature could replace laravels `$browser->element(...)` functionality too

**Comment by ebebbington** [comment]
There is a proper fix in the works, but for the time being, a temp fix is in place (see https://github.com/denoland/deno/pull/9288)

When deno releases with the fix, we can remove this temp fix.

Though this means that, is a user wants to use Rhum + the evaluatePage() method, it won't work for them unless they have this temp fix, as opposed to typescript itself fixing an upstream issue (which they probably wont)

Does anyone object tot his temp fix?

**Comment by ebebbington** [comment]
Docs feature will have to be updated, to use the new lib `dom.asynciterable` in the `tsconfig.json`

**Comment by ebebbington** [comment]
> Docs feature will have to be updated, to use the new lib `dom.asynciterable` in the `tsconfig.json`

Nevermind it doesn't, we encountered this issue because. we are using Rhum, and rhum was using std/async/pool.ts, which was the cause of the problem

**Review by ebebbington — APPROVED** [review]
@saragee3 im happy with your changes :)

**Comment by saragee3** [comment]
feel free to ship it! sorry, been M I A.

**Comment by ebebbington** [comment]
> @ebebbington feel free to ship it! sorry, been M I A.

No problem! We can ship it once the docs MR is approved and merged, then we can release this feature and deploy the docs at the same time

**Pull request merged**
